students/YingGuo/lession07/html_render.py

```diff
@@ -0,0 +1,162 @@
+            out_file.write(cur_ind)
+        else:
+            pass
+
+        for content in self.content:
+            out_file.write("<{}>\n".format(self.tag_name))
+
+            if cur_ind:
+                out_file.write(cur_ind)
+            else:
+                pass
+            #
+            out_file.write(self.indent)
+            try:
+                content.render(out_file)
+            except AttributeError:
+                out_file.write(content)
+            out_file.write("\n")
+            
+            out_file.write("</{}>\n".format(self.tag_name))
+
+
+#a subclass of Element with tag body
+class Body(Element):
+    tag_name = "body"
+    indent = "    "*2
+
+#subclass for paragraph
+class P(Element):
+    tag_name = "p"
+    indent = "    "*3
+
+    def render(self, out_file, cur_ind=""):
+        if cur_ind:
+            out_file.write(cur_ind)
+        else:
+            pass
+
+        for content in self.content:
+            if self.kwargs is not None:
+                open_tag = ["<{}".format(self.tag_name)]
+                for x in self.kwargs:
+                    open_tag.append(x)
+                    open_tag.append("=")
+                    open_tag.append('"{}"'.format(self.kwargs[x]))
+                    open_tag.append(" ")
+                open_tag.append(">\n")
+                out_file.write("".join(open_tag))
+            else:
+                pass
+
+            out_file.write(self.indent)
+
+            try:
+                content.render(out_file)
+            except AttributeError:
+                out_file.write(content)
+            out_file.write("\n")
+            out_file.write("</{}>\n".format(self.tag_name))
+
+class Html(Element):
+    tag_name = "html"
+
+class Title(Element):
+    tag_name = "title"
+    def render(self, out_file):
+        for content in self.content:
+            out_file.write("<{}>".format(self.tag_name))
+            try:
+                content.render(out_file)
+            except AttributeError:
+                out_file.write(content)
+            out_file.write("</{}>\n".format(self.tag_name))
+
+class OneLineTag(Element):
+    def render(self, out_file, cur_ind=""):
+            out_file.write("<{}>".format(self.tag_name))
+            out_file.write(self.content[0])
+            out_file.write("</{}>\n".format(self.tag_name))
+    #change the test file with import pytest
+    def append(self, content):
+        raise NotImplementedError
+
+class SelfClosingTag(Element):
+    #somethinkg like <hr width="400" />
+    def render(self, out_file, cur_ind=""):
+        open_tag = ["<{} ".format(self.tag_name)]
+        if self.kwargs:
+            for x in self.kwargs:
+                open_tag.append(x)
+                open_tag.append("=")
+                open_tag.append('"{}"'.format(self.kwargs[x]))
+                out_file.write("".join(open_tag))
+                out_file.write(" ")
+        else:
+            out_file.write("".join(open_tag))
```

Comment:
You had to duplicate the same attribute-rendering code here as well, because you didn't put the earlier code in the `Element` class. The key point is to avoid this kind of code duplication by utilizing inheritance and reusing the inherited methods as is.